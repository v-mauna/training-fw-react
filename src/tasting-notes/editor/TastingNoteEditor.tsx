import React, { useEffect, useState } from 'react';
import {
  IonHeader,
  IonContent,
  IonFooter,
  IonTitle,
  IonToolbar,
  IonButton,
  IonButtons,
  IonIcon,
  IonInput,
  IonItem,
  IonLabel,
  IonSelect,
  IonSelectOption,
  IonTextarea,
} from '@ionic/react';
import { close } from 'ionicons/icons';
import { Controller, useForm } from 'react-hook-form';
import { TastingNote, Tea } from '../../shared/models';
import { Rating } from '../../shared/components';
import { useTea } from '../../tea/useTea';
import { useTastingNotes } from '../useTastingNotes';

interface TastingNoteEditorProps {
  onDismiss: (opts: { refresh: boolean }) => void;
  note?: TastingNote;
}

const TastingNoteEditor: React.FC<TastingNoteEditorProps> = ({
  onDismiss,
  note = undefined,
}) => {
  const { handleSubmit, control, formState } = useForm<TastingNote>({
    mode: 'onChange',
  });
  const [teas, setTeas] = useState<Array<Tea>>([]);
  const { getTeas } = useTea();
  const { saveNote } = useTastingNotes();

  useEffect(() => {
    const init = async () => {
      const teas = await getTeas();
      setTeas(teas);
    };
    init();
  }, [getTeas]);

  const save = async (data: TastingNote) => {
    if (note?.id) data.id = note.id;
    await saveNote(data);
    onDismiss({ refresh: true });
  };

  return (
    <>
      <IonHeader>
        <IonToolbar>
          <IonTitle>{!note && 'Add New'} Tasting Note</IonTitle>
          <IonButtons slot="primary">
            <IonButton
              id="cancel-button"
              onClick={() => onDismiss({ refresh: false })}
            >
              <IonIcon slot="icon-only" icon={close} />
            </IonButton>
          </IonButtons>
        </IonToolbar>
      </IonHeader>

      <IonContent>
        <form>
          <IonItem>
            <IonLabel position="floating">Brand</IonLabel>
            <Controller
              render={({ onChange, value }) => (
                <IonInput
                  id="brand-input"
                  onIonChange={(e: any) => onChange(e.detail.value!)}
                  value={value}
                />
              )}
              control={control}
              name="brand"
              rules={{ required: true }}
              defaultValue={note?.brand || ''}
            />
          </IonItem>
          <IonItem>
            <IonLabel position="floating">Name</IonLabel>
            <Controller
              render={({ onChange, value }) => (
                <IonInput
                  id="name-input"
                  onIonChange={(e: any) => onChange(e.detail.value!)}
                  value={value}
                />
              )}
              control={control}
              name="name"
              rules={{ required: true }}
              defaultValue={note?.name || ''}
            />
          </IonItem>
          <IonItem>
            <IonLabel>Category</IonLabel>
            <Controller
              render={({ onChange, value }) => (
                <>
                  {teas.length && (
                    <IonSelect
                      onIonChange={(e: any) => onChange(e.detail.value!)}
                      value={value}
                    >
                      {teas.map((tea: Tea) => (
                        <IonSelectOption key={tea.id} value={tea.id}>
                          {tea.name}
                        </IonSelectOption>
                      ))}
                    </IonSelect>
                  )}
                </>
              )}
              control={control}
              name="teaCategoryId"
              rules={{ required: true }}
              defaultValue={note?.teaCategoryId || 1}
            />
          </IonItem>
          <IonItem>
            <IonLabel>Rating</IonLabel>
            <Controller
              render={({ onChange, value }) => (
                <Rating onRatingChange={onChange} initialRating={value} />
              )}
              control={control}
              name="rating"
              rules={{ required: true }}
              defaultValue={note?.rating || 0}
            />
          </IonItem>
          <IonItem>
            <IonLabel position="floating">Notes</IonLabel>
            <Controller
              render={({ onChange, value }) => (
                <IonTextarea
                  id="notes-input"
                  onIonChange={(e: any) => onChange(e.detail.value!)}
                  rows={5}
                  value={value}
                />
              )}
              control={control}
              name="notes"
              rules={{ required: true }}
              defaultValue={note?.notes || ''}
            />
          </IonItem>
        </form>
      </IonContent>

      <IonFooter>
        <IonToolbar>
          <IonButton
            type="submit"
            disabled={!formState.isValid}
            expand="full"
            onClick={handleSubmit(data => save(data))}
          >
            {note ? 'Update' : 'Add'}
          </IonButton>
        </IonToolbar>
      </IonFooter>
    </>
  );
};
export default TastingNoteEditor;
