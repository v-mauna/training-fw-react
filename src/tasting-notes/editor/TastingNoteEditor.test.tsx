import React from 'react';
import { render, wait, waitForElement } from '@testing-library/react';
import { ionFireEvent as fireEvent } from '@ionic/react-test-utils';
import { cleanup } from '@testing-library/react-hooks';
import TastingNoteEditor from './TastingNoteEditor';
jest.mock('../../tea/useTea', () => ({
  useTea: () => ({
    getTeas: jest.fn(() => Promise.resolve(mockTeas)),
  }),
}));
jest.mock('../useTastingNotes', () => ({
  useTastingNotes: () => ({
    saveNote: mockSaveNote,
  }),
}));

let mockSaveNote = jest.fn(() => Promise.resolve());

const mockNote = {
  id: 73,
  brand: 'Lipton',
  name: 'Yellow Label',
  teaCategoryId: 1,
  rating: 1,
  notes: 'ick',
};

const mockTeas = [
  {
    id: 7,
    name: 'White',
    image: 'assets/img/white.jpg',
    description: 'White tea description.',
    rating: 5,
  },
  {
    id: 8,
    name: 'Yellow',
    image: 'assets/img/yellow.jpg',
    description: 'Yellow tea description.',
    rating: 3,
  },
];

describe('<TastingNoteEditor />', () => {
  let component: any;
  let mockDismiss = jest.fn();

  beforeEach(() => {
    component = <TastingNoteEditor onDismiss={mockDismiss} />;
    mockSaveNote = jest.fn(() => Promise.resolve());
  });

  describe('initialization', () => {
    it('binds the tea select', async () => {
      const { container } = render(component);
      const options = await waitForElement(
        () => container.querySelector('ion-select')!.children,
      );
      expect(options.length).toEqual(2);
      expect(options[0].textContent).toEqual('White');
      expect(options[1].textContent).toEqual('Yellow');
    });
  });

  describe('cancel button', () => {
    it('calls the dismiss function', async () => {
      const { container } = render(component);
      const button = await waitForElement(
        () => container.querySelector('#cancel-button')!,
      );
      fireEvent.click(button);
      expect(mockDismiss).toHaveBeenCalledTimes(1);
    });
  });

  describe('save', () => {
    it('renders consistently', async () => {
      const { asFragment } = render(component);
      await wait(() => expect(asFragment()).toMatchSnapshot());
    });

    it('saves the note', async () => {
      const expected = { ...mockNote };
      delete expected.id;
      const { container, getByLabelText } = render(component);
      const [brand, name, rating, notes, submit] = await waitForElement(() => [
        container.querySelector('#brand-input')!,
        container.querySelector('#name-input')!,
        getByLabelText(/Rate 1 stars/),
        container.querySelector('#notes-input')!,
        container.querySelector('[type="submit"]')! as HTMLIonButtonElement,
      ]);
      await wait(() => {
        fireEvent.ionChange(brand, mockNote.brand);
        fireEvent.ionChange(name, mockNote.name);
        fireEvent.click(rating);
        fireEvent.ionChange(notes, mockNote.notes);
        fireEvent.click(submit);
      });
      expect(mockSaveNote).toHaveBeenCalledWith(expected);
      expect(mockSaveNote).toHaveBeenCalledTimes(1);
    });

    it('has the add title', async () => {
      const { container } = render(component);
      await wait(() =>
        expect(container).toHaveTextContent(/Add New Tasting Note/),
      );
    });

    it('has the add button label', async () => {
      const { container } = render(component);
      const submit = await waitForElement(
        () => container.querySelector('[type="submit"]')!,
      );
      expect((submit as HTMLIonButtonElement).textContent).toEqual('Add');
    });
  });

  describe('update', () => {
    beforeEach(() => {
      component = <TastingNoteEditor onDismiss={mockDismiss} note={mockNote} />;
      mockSaveNote = jest.fn(() => Promise.resolve());
    });

    it('renders consistently', async () => {
      const { asFragment } = render(component);
      await wait(() => expect(asFragment()).toMatchSnapshot());
    });

    it('sets the properties', async () => {
      const { container } = render(component);
      const [brand, name, notes] = await waitForElement(() => [
        container.querySelector('#brand-input')!,
        container.querySelector('#name-input')!,
        container.querySelector('#notes-input')!,
      ]);
      expect(brand.getAttribute('value')).toEqual(mockNote.brand);
      expect(name.getAttribute('value')).toEqual(mockNote.name);
      expect(notes.getAttribute('value')).toEqual(mockNote.notes);
    });

    it('updates the data', async () => {
      const expected = { ...mockNote };
      expected.notes = "It's not good";
      const { container, getByLabelText } = render(component);
      const [brand, name, rating, notes, submit] = await waitForElement(() => [
        container.querySelector('#brand-input')!,
        container.querySelector('#name-input')!,
        getByLabelText(/Rate 1 stars/),
        container.querySelector('#notes-input')!,
        container.querySelector('[type="submit"]')! as HTMLIonButtonElement,
      ]);
      await wait(() => {
        fireEvent.ionChange(brand, mockNote.brand);
        fireEvent.ionChange(name, mockNote.name);
        fireEvent.click(rating);
        fireEvent.ionChange(notes, expected.notes);
        fireEvent.click(submit);
      });
      expect(mockSaveNote).toHaveBeenCalledWith(expected);
      expect(mockSaveNote).toHaveBeenCalledTimes(1);
    });

    it('has the update title', async () => {
      const { container } = render(component);
      await wait(() => expect(container).toHaveTextContent(/Tasting Note/));
    });

    it('has the update button label', async () => {
      const { container } = render(component);
      const submit = await waitForElement(
        () => container.querySelector('[type="submit"]')!,
      );
      expect((submit as HTMLIonButtonElement).textContent).toEqual('Update');
    });
  });

  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });
});
