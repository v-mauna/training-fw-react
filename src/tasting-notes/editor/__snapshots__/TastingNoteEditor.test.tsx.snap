// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TastingNoteEditor /> save renders consistently 1`] = `
<DocumentFragment>
  <ion-header>
    <ion-toolbar>
      <ion-title>
        Add New Tasting Note
      </ion-title>
      <ion-buttons
        slot="primary"
      >
        <ion-button
          id="cancel-button"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z'/></svg>"
            slot="icon-only"
          />
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header>
  <ion-content>
    <form>
      <ion-item>
        <ion-label
          position="floating"
        >
          Brand
        </ion-label>
        <ion-input
          id="brand-input"
          value=""
        />
      </ion-item>
      <ion-item>
        <ion-label
          position="floating"
        >
          Name
        </ion-label>
        <ion-input
          id="name-input"
          value=""
        />
      </ion-item>
      <ion-item>
        <ion-label>
          Category
        </ion-label>
        <ion-select>
          <ion-select-option>
            White
          </ion-select-option>
          <ion-select-option>
            Yellow
          </ion-select-option>
        </ion-select>
      </ion-item>
      <ion-item>
        <ion-label>
          Rating
        </ion-label>
        <div
          class="rating"
          style="opacity: 1;"
        >
          <ion-icon
            aria-label="Rate 1 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 2 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 3 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 4 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 5 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
        </div>
      </ion-item>
      <ion-item>
        <ion-label
          position="floating"
        >
          Notes
        </ion-label>
        <ion-textarea
          id="notes-input"
          value=""
        />
      </ion-item>
    </form>
  </ion-content>
  <ion-footer>
    <ion-toolbar>
      <ion-button
        expand="full"
        type="submit"
      >
        Add
      </ion-button>
    </ion-toolbar>
  </ion-footer>
</DocumentFragment>
`;

exports[`<TastingNoteEditor /> update renders consistently 1`] = `
<DocumentFragment>
  <ion-header>
    <ion-toolbar>
      <ion-title>
         Tasting Note
      </ion-title>
      <ion-buttons
        slot="primary"
      >
        <ion-button
          id="cancel-button"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z'/></svg>"
            slot="icon-only"
          />
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header>
  <ion-content>
    <form>
      <ion-item>
        <ion-label
          position="floating"
        >
          Brand
        </ion-label>
        <ion-input
          id="brand-input"
          value="Lipton"
        />
      </ion-item>
      <ion-item>
        <ion-label
          position="floating"
        >
          Name
        </ion-label>
        <ion-input
          id="name-input"
          value="Yellow Label"
        />
      </ion-item>
      <ion-item>
        <ion-label>
          Category
        </ion-label>
        <ion-select>
          <ion-select-option>
            White
          </ion-select-option>
          <ion-select-option>
            Yellow
          </ion-select-option>
        </ion-select>
      </ion-item>
      <ion-item>
        <ion-label>
          Rating
        </ion-label>
        <div
          class="rating"
          style="opacity: 1;"
        >
          <ion-icon
            aria-label="Rate 1 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M394 480a16 16 0 01-9.39-3L256 383.76 127.39 477a16 16 0 01-24.55-18.08L153 310.35 23 221.2a16 16 0 019-29.2h160.38l48.4-148.95a16 16 0 0130.44 0l48.4 149H480a16 16 0 019.05 29.2L359 310.35l50.13 148.53A16 16 0 01394 480z'/></svg>"
          />
          <ion-icon
            aria-label="Rate 2 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 3 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 4 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
          <ion-icon
            aria-label="Rate 5 stars"
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M480 208H308L256 48l-52 160H32l140 96-54 160 138-100 138 100-54-160z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
        </div>
      </ion-item>
      <ion-item>
        <ion-label
          position="floating"
        >
          Notes
        </ion-label>
        <ion-textarea
          id="notes-input"
          value="ick"
        />
      </ion-item>
    </form>
  </ion-content>
  <ion-footer>
    <ion-toolbar>
      <ion-button
        expand="full"
        type="submit"
      >
        Update
      </ion-button>
    </ion-toolbar>
  </ion-footer>
</DocumentFragment>
`;
